### PR TITLE
[DeepSeek-V4] Add mhc_fused_post_pre kernel

### PR DESCRIPTION
**Pull request by JoeLee314**
# [DeepSeek-V4] Add fused mHC post/pre kernel

## Motivation

This PR optimizes the DeepSeek-V4 mHC path by fusing the boundary between one mHC post step and the next mHC pre step.

The approach is adapted from TRTLLM. The main idea is to avoid launching separate kernels for the latency-sensitive decode path:

```text
mhc_post -> pre-norm GEMM -> mhc_pre finalization
```

The fused path keeps the existing numerically sensitive `mhc_pre_big_fuse(_with_norm)` finalization stage, but removes the separate post/pre scheduling boundary.

For small token batches, a TileLang scalar-FMA kernel computes the following work in one launch:

```text
previous hc_post + bf16 residual materialization + pre-norm GEMM partials + RMS square-sum partials
```

For larger token batches, the implementation keeps the existing high-throughput TileLang `mhc_post` + DeepGEMM decomposition.

The optimization is guarded by a new env var:

```bash
SGLANG_OPT_FUSE_MHC_POST_PRE=1
```

It intentionally depends on the existing TileLang mHC switches:

```bash
SGLANG_OPT_USE_TILELANG_MHC_PRE=1
SGLANG_OPT_USE_TILELANG_MHC_POST=1
```

because the fused implementation directly reuses the TileLang mHC tensor layouts and the `mhc_pre_big_fuse(_with_norm)` finalization kernels.

## Performance Test
total_throughput increased by 3.35%
### Performance summary

image

## Accuracy / Correctness Tests
GSM8K score: 0.975

### `python/sglang/srt/environ.py`

Add:

```python
SGLANG_OPT_FUSE_MHC_POST_PRE = EnvBool(False)
```

### `python/sglang/srt/layers/mhc.py`

- Add `mhc_fused_post_pre_fma_tilelang`, a small-token fused TileLang kernel.
- Add `mhc_fused_post_pre`, which dispatches:
  - small batches to the fused FMA kernel;
  - larger batches to the existing TileLang `mhc_post` + DeepGEMM path.
- Preserve the existing `mhc_pre_big_fuse(_with_norm)` finalization stage.
- Handle zero-token DP/EP ranks with correctly typed empty tensors.

### `python/sglang/srt/models/deepseek_v4.py`

- Add fused mHC enable check cached in `__init__`.
- Fuse:
  - cross-layer previous FFN `hc_post` + next attention `hc_pre`;
  - within-layer attention `hc_post` + FFN `hc_pre`.
- Defer the final FFN `hc_post` and close it at model tail.
- Cache bf16 RMSNorm weights for the fused path to avoid per-forward casts.

### `python/sglang/srt/models/deepseek_v4_nextn.py`

- Adapt NextN decoder usage to the decoder layer returning deferred fused state.

### `tests/kernels/test_mhc_kernels.py`

- Add coverage for `mhc_fused_post_pre` vs the unfused reference sequence.

## Checklist

- [x] Added fused mHC post/pre boundary optimization.
- [x] Added env guard `SGLANG_OPT_FUSE_MHC_POST_PRE`.
- [x] Kept the fused path dependent on TileLang mHC pre/post kernels.
- [x] Added kernel unit test for fused vs unfused mHC.
- [x] Verified operator-level precision.
- [x] Verified short deterministic sanity prompts.
- [x] Verified GSM8K-200 end-to-end accuracy.
- [x] Verified `bench_serving` performance improvement on DeepSeek-V4-Flash.

---
### CI States

Latest PR Test (Base): :x: [Run #26549213709](https://github.com/sgl-project/sglang/actions/runs/26549213709)
Latest PR Test (Extra): :x: [Run #26549213622](https://github.com/sgl-project/sglang/actions/runs/26549213622)

### CODE REVIEW

File: tests/kernels/test_mhc_kernels.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+@pytest.mark.parametrize("num_tokens", [0, 1, 8, 17, 32])
+@pytest.mark.parametrize("use_norm", [False, True])
+def test_mhc_fused_post_pre_matches_unfused(monkeypatch, num_tokens, use_norm):
+    if not torch.cuda.is_available():
+        pytest.skip("CUDA is required for TileLang mHC kernels")
+
+    monkeypatch.setattr(mhc, "is_dsa_prefill_cp_round_robin_split", lambda: False)
+    torch.manual_seed(0)
+    device = torch.device("cuda")
+    hc_mult = 2
```
Comment (yhyang201) [inline tests/kernels/test_mhc_kernels.py]: Could we change this to `hc_mult=4` to match 4's actual config?

File: tests/kernels/test_mhc_kernels.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
+import pytest
+import torch
+
+import sglang.srt.layers.mhc as mhc
+from sglang.srt.layers.mhc import mhc_fused_post_pre, mhc_post, mhc_pre
+
+
+@pytest.mark.parametrize("num_tokens", [0, 1, 8, 17, 32])
```
Comment (yhyang201) [inline tests/kernels/test_mhc_kernels.py]: Can we add a batch size of 64 to test the correctness of the original (unfused large-batch) path?

File: python/sglang/srt/layers/mhc.py [hunk 3]
```diff
@@ -914,3 +914,500 @@ def mhc_post(
...
+            post_layer_mix.view(num_tokens, hc_mult),
+            x_flat,
+            fn.view(hc_mult3, hc_mult, hidden_size),
+            gemm_out_mul,
+            gemm_out_sqrsum,
+            residual_cur,
+            hc_mult,
+            hidden_size,
+            hc_mult3,
+            tile_n=tile_n,
```
Comment (yhyang201) [inline python/sglang/srt/layers/mhc.py]: It looks like `mhc_fused_post_pre_fma_tilelang` doesn’t have the parameter `tile_n`?

File: tests/kernels/test_mhc_kernels.py (hunk 1 above)
Comment (JoeLee314) [reply tests/kernels/test_mhc_kernels.py]: Done

File: tests/kernels/test_mhc_kernels.py (hunk 2 above)
Comment (JoeLee314) [reply tests/kernels/test_mhc_kernels.py]: Done

File: python/sglang/srt/layers/mhc.py (hunk 3 above)
Comment (JoeLee314) [reply python/sglang/srt/layers/mhc.py]: You’re right! `tile_n` should be passed to `tile_mix_outputs`, I missed that during the refactor. 
Thanks for catching it, fixed now.

File: tests/kernels/test_mhc_kernels.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+@pytest.mark.parametrize("use_norm", [False, True])
+def test_mhc_fused_post_pre_matches_unfused(monkeypatch, num_tokens, use_norm):
+    if not torch.cuda.is_available():
+        pytest.skip("CUDA is required for TileLang mHC kernels")
+
+    monkeypatch.setattr(mhc, "is_dsa_prefill_cp_round_robin_split", lambda: False)
+    torch.manual_seed(0)
+    device = torch.device("cuda")
+    hc_mult = 4
+    hidden_size = 4096
```
Comment (yhyang201) [inline tests/kernels/test_mhc_kernels.py]: can we parameterize `hidden_size` and add 7168 (v4pro)?

File: python/sglang/srt/layers/mhc.py [hunk 5]
```diff
@@ -914,3 +914,500 @@ def mhc_post(
...
+
+
+@tilelang.jit(
+    pass_configs={
+        tilelang.PassConfigKey.TL_DISABLE_WARP_SPECIALIZED: True,
+        tilelang.PassConfigKey.TL_DISABLE_TMA_LOWER: True,
+        tilelang.PassConfigKey.TL_PTXAS_REGISTER_USAGE_LEVEL: 10,
+    },
+)
+def mhc_fused_post_pre_fma_tilelang(
```
Comment (yhyang201) [inline python/sglang/srt/layers/mhc.py:908]: Can we add warmup for the fused kernel in `prewarm_mhc_token_counts`?

Comment (JoeLee314) [reply python/sglang/srt/layers/mhc.py]: Done

File: tests/kernels/test_mhc_kernels.py (hunk 4 above)
Comment (JoeLee314) [reply tests/kernels/test_mhc_kernels.py]: Done

File: python/sglang/srt/layers/mhc.py (hunk 5 above)
Comment (JoeLee314) [reply python/sglang/srt/layers/mhc.py:908]: Done

**Review by yhyang201 — APPROVED** [review]
